File: 1628-count-submatrices-with-all-ones/1628-count-submatrices-with-all-ones.cpp
class Solution {
public:
    int numSubmat(vector<vector<int>>& mat) {
        int res = 0;
        vector<vector<int>> widths(mat.size(),
                               vector<int>(mat[0].size(), 0));
        for (int i = 0; i < mat.size(); i++) {
            for (int j = 0; j < mat[i].size(); j++) {
                if(mat[i][j] == 1)
                {
                    if(j == 0)
                    {
                        widths[i][j] = 1;
                    }
                    else
                    {
                        widths[i][j] = widths[i][j - 1] + 1;
                    }
                }
            }
            for (int j = 0; j < mat[i].size(); j++) {
                int minWidth = widths[i][j];
                for(int k = i; k >= 0; k--)
                {
                    minWidth = min(minWidth, widths[k][j]);
                    if(minWidth == 0)
                    {
                        break;
                    }
                    res += minWidth;
                }
            }
        }

        return res;
    }
};

File: 1628-count-submatrices-with-all-ones/1628-count-submatrices-with-all-ones.py
class Solution:
    def numSubmat(self, mat: List[List[int]]) -> int:
        res = 0
        widths = []
        for row in mat:
            width = []
            for cell in row:
                if cell == 1:
                    if len(width) == 0:
                        width.append(1)
                    else:
                        width.append(width[-1] + 1)
                else:
                    width.append(0)
            widths.append(width)
            for j in range(len(width)):
                minWidth = width[j]
                for k in range(len(widths) - 1, -1, -1):
                    minWidth = min(minWidth, widths[k][j])
                    if minWidth == 0:
                        break;
                    res += minWidth
        return res
